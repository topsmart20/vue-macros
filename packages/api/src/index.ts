export * from '@vue-macros/common'

export * from './vue'
export * from './resolve'
export * from './ts'
export * from './utils'
