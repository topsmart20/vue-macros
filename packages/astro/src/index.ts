import { type AstroIntegration, type ViteUserConfig } from 'astro'
import { type Options, resolveOptions } from 'unplugin-vue-macros'
import { type Plugin } from 'vite'
import VueMacros from 'unplugin-vue-macros/vite'

export type VueMacrosOptions = Options

function findPluginAndRemove(
  name: string,
  plugins: ViteUserConfig['plugins']
): Plugin | undefined {
  const idx = plugins!.findIndex(
    (plugin) => plugin && 'name' in plugin && plugin.name === name
  )
  if (idx === -1) return
  const plugin = plugins![idx]
  plugins!.splice(idx, 1)
  return plugin as any
}

export default function (options?: VueMacrosOptions): AstroIntegration {
  return {
    name: '@vue-macros/astro',
    hooks: {
      'astro:config:setup': ({ config }) => {
        const resolvedOptions = resolveOptions(options ?? {})
        const vue = findPluginAndRemove('vite:vue', config.vite.plugins)
        const vueJsx = findPluginAndRemove('vite:vue-jsx', config.vite.plugins)

        config.vite.plugins?.push(
          VueMacros({
            ...resolvedOptions,
            plugins: {
              vue,
              vueJsx,
            },
          })
        )
      },
    },
  }
}
