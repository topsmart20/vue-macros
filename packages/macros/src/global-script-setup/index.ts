import path from 'node:path'
import {
  DEFINE_SETUP_COMPONENT,
  MagicString,
  babelParse,
  getTransformResult,
} from '@vue-macros/common'
import { walk } from 'estree-walker'
import { GLOBAL_SCRIPT_SETUP_ID_PREFIX } from './constants'
import type { CallExpression, Function, Node } from '@babel/types'

export * from './constants'

interface NodeContext {
  code: string
  body: string
  node: CallExpression
}
export type GlobalScriptSetupContext = Record<string, NodeContext[]>

export const transformGlobalScriptSetup = (
  code: string,
  id: string,
  ctx: GlobalScriptSetupContext
) => {
  const program = babelParse(code, path.extname(id).replace(/^\./, ''))
  const nodes: CallExpression[] = []
  walk(program, {
    enter(node: Node) {
      if (
        node.type === 'CallExpression' &&
        node.callee.type === 'Identifier' &&
        node.callee.name === DEFINE_SETUP_COMPONENT
      ) {
        nodes.push(node)
      }
    },
  })

  if (nodes.length === 0) return

  const s = new MagicString(code)
  const nodeContexts: NodeContext[] = nodes.map((node, i) => {
    if (
      !['FunctionExpression', 'ArrowFunctionExpression'].includes(
        node.arguments[0].type
      )
    )
      throw new SyntaxError(
        `${DEFINE_SETUP_COMPONENT}: Invalid setup component definition`
      )

    const importName = `setupComponent_${i}`
    s.overwrite(node.start!, node.end!, importName)
    s.prepend(
      `import ${importName} from '${GLOBAL_SCRIPT_SETUP_ID_PREFIX}${id}:${i}.vue'\n`
    )
    const body: Node = (node.arguments[0] as Function).body
    let bodyStart = body.start!
    let bodyEnd = body.end!
    if (body.type === 'BlockStatement') {
      bodyStart++
      bodyEnd--
    }
    return {
      code: code.slice(node.start!, node.end!),
      body: code.slice(bodyStart, bodyEnd),
      node,
    }
  })
  ctx[id] = nodeContexts

  return getTransformResult(s, id)
}

export const loadGlobalScriptSetup = (
  id: string,
  ctx: GlobalScriptSetupContext
) => {
  const [, file, i] = id.split(':')
  const index = +i.replace('.vue', '')
  if (!ctx[file]?.[index]) return

  const { body } = ctx[file][index]
  const lang = path.extname(file).replace(/^\./, '')
  const code = `<script setup${lang ? ` lang="${lang}"` : ''}>
${body}
</script>`
  return code
}
