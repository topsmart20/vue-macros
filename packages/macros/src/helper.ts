export function emitHelper(
  emitFn: (evt: string, ...args: any[]) => void,
  key: string,
  value: unknown,
  ...args: [returnValue?: unknown]
) {
  emitFn(key, value)
  return args.length > 0 ? args[0] : value
}
