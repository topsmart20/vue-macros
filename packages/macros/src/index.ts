import { type UnpluginInstance } from 'unplugin'
import {
  type OptionsPlugin,
  type Plugin,
  type PluginType,
  createCombinePlugin,
} from 'unplugin-combine'

import { detectVueVersion } from '@vue-macros/common'
import { Devtools } from '@vue-macros/devtools'

import VueBetterDefine, {
  type Options as OptionsBetterDefine,
} from '@vue-macros/better-define'
import VueChainCall, {
  type Options as OptionsChainCall,
} from '@vue-macros/chain-call'
import VueDefineEmit, {
  type Options as OptionsDefineEmit,
} from '@vue-macros/define-emit'
import VueDefineModels, {
  type Options as OptionsDefineModels,
} from '@vue-macros/define-models'
import VueDefineOptions, {
  type Options as OptionsDefineOptions,
} from 'unplugin-vue-define-options'
import VueDefineProp, {
  type Options as OptionsDefineProp,
} from '@vue-macros/define-prop'
import VueDefineProps, {
  type Options as OptionsDefineProps,
} from '@vue-macros/define-props'
import VueDefinePropsRefs, {
  type Options as OptionsDefinePropsRefs,
} from '@vue-macros/define-props-refs'
import VueDefineRender, {
  type Options as OptionsDefineRender,
} from '@vue-macros/define-render'
import VueDefineSlots, {
  type Options as OptionsDefineSlots,
} from '@vue-macros/define-slots'
import VueExportExpose, {
  type Options as OptionsExportExpose,
} from '@vue-macros/export-expose'
import VueExportProps, {
  type Options as OptionsExportProps,
} from '@vue-macros/export-props'
import VueHoistStatic, {
  type Options as OptionsHoistStatic,
} from '@vue-macros/hoist-static'
import VueJsxDirective, {
  type Options as OptionsJsxDirective,
} from '@vue-macros/jsx-directive'
import VueNamedTemplate, {
  type Options as OptionsNamedTemplate,
} from '@vue-macros/named-template'
import VueReactivityTransform, {
  type Options as OptionsReactivityTransform,
} from '@vue-macros/reactivity-transform'
import VueSetupBlock, {
  type Options as OptionsSetupBlock,
} from '@vue-macros/setup-block'
import VueSetupComponent, {
  type Options as OptionsSetupComponent,
} from '@vue-macros/setup-component'
import VueSetupSFC, {
  type Options as OptionsSetupSFC,
} from '@vue-macros/setup-sfc'
import VueShortEmits, {
  type Options as OptionsShortEmits,
} from '@vue-macros/short-emits'

export interface FeatureOptionsMap {
  betterDefine: OptionsBetterDefine
  chainCall: OptionsChainCall
  defineEmit: OptionsDefineEmit
  defineModels: OptionsDefineModels
  defineOptions: OptionsDefineOptions
  defineProp: OptionsDefineProp
  defineProps: OptionsDefineProps
  definePropsRefs: OptionsDefinePropsRefs
  defineRender: OptionsDefineRender
  defineSlots: OptionsDefineSlots
  exportExpose: OptionsExportExpose
  exportProps: OptionsExportProps
  hoistStatic: OptionsHoistStatic
  jsxDirective: OptionsJsxDirective
  namedTemplate: OptionsNamedTemplate
  reactivityTransform: OptionsReactivityTransform
  setupBlock: OptionsSetupBlock
  setupComponent: OptionsSetupComponent
  setupSFC: OptionsSetupSFC
  shortEmits: OptionsShortEmits
}
export type FeatureName = keyof FeatureOptionsMap
export type FeatureOptions = FeatureOptionsMap[FeatureName]

export interface OptionsCommon {
  root?: string
  version?: number
  isProduction?: boolean
  plugins?: {
    vue?: any
    vueJsx?: any
  }
  /** @internal */
  nuxtContext?: {
    isClient?: boolean
  }
}

type OptionalSubOptions<T> = boolean | Omit<T, keyof OptionsCommon> | undefined

export type Options = OptionsCommon & {
  [K in FeatureName]?: OptionalSubOptions<FeatureOptionsMap[K]>
}

export type OptionsResolved = Required<OptionsCommon> & {
  [K in FeatureName]: false | FeatureOptionsMap[K]
}

export function resolveOptions({
  root,
  version,
  plugins,
  isProduction,
  nuxtContext,

  betterDefine,
  chainCall,
  defineEmit,
  defineModels,
  defineOptions,
  defineProp,
  defineProps,
  definePropsRefs,
  defineRender,
  defineSlots,
  exportExpose,
  exportProps,
  hoistStatic,
  jsxDirective,
  namedTemplate,
  reactivityTransform,
  setupBlock,
  setupComponent,
  setupSFC,
  shortEmits,
}: Options): OptionsResolved {
  function resolveSubOptions<K extends FeatureName>(
    options: OptionalSubOptions<FeatureOptionsMap[K]>,
    commonOptions: Required<
      Pick<OptionsCommon, keyof OptionsCommon & keyof FeatureOptionsMap[K]>
    >,
    defaultEnabled = true
  ): FeatureOptionsMap[K] | false {
    options = options ?? defaultEnabled
    if (!options) return false
    return { ...(options === true ? {} : options), ...commonOptions }
  }

  root = root || process.cwd()
  version = version || detectVueVersion(root)
  isProduction = isProduction ?? process.env.NODE_ENV === 'production'

  return {
    plugins: plugins || {},
    root,
    version,
    isProduction,
    nuxtContext: nuxtContext || {},

    betterDefine: resolveSubOptions<'betterDefine'>(betterDefine, {
      version,
      isProduction,
    }),
    chainCall: resolveSubOptions<'chainCall'>(chainCall, { version }),
    defineEmit: resolveSubOptions<'defineEmit'>(defineEmit, {
      isProduction,
      version,
    }),
    defineModels: resolveSubOptions<'defineModels'>(defineModels, { version }),
    defineOptions: resolveSubOptions<'defineOptions'>(
      defineOptions,
      { version },
      version < 3.3
    ),
    defineProp: resolveSubOptions<'defineProp'>(defineProp, {
      isProduction,
      version,
    }),
    defineProps: resolveSubOptions<'defineProps'>(defineProps, { version }),
    definePropsRefs: resolveSubOptions<'definePropsRefs'>(definePropsRefs, {
      version,
    }),
    defineRender: resolveSubOptions<'defineRender'>(defineRender, { version }),
    defineSlots: resolveSubOptions<'defineSlots'>(
      defineSlots,
      { version },
      version < 3.3
    ),
    exportExpose: resolveSubOptions<'exportExpose'>(
      exportExpose,
      { version },
      false
    ),
    exportProps: resolveSubOptions<'exportProps'>(
      exportProps,
      { version },
      false
    ),
    jsxDirective: resolveSubOptions<'jsxDirective'>(jsxDirective, {
      version,
    }),
    hoistStatic: resolveSubOptions<'hoistStatic'>(hoistStatic, { version }),
    namedTemplate: resolveSubOptions<'namedTemplate'>(namedTemplate, {
      version,
    }),
    reactivityTransform: resolveSubOptions<'reactivityTransform'>(
      reactivityTransform,
      { version }
    ),
    setupBlock: resolveSubOptions<'setupBlock'>(setupBlock, { version }, false),
    setupComponent: resolveSubOptions<'setupComponent'>(setupComponent, {
      version,
      root,
    }),
    setupSFC: resolveSubOptions<'setupSFC'>(setupSFC, { version }, false),
    shortEmits: resolveSubOptions<'shortEmits'>(
      shortEmits,
      { version },
      version < 3.3
    ),
  }
}

function resolvePlugin(
  unplugin: UnpluginInstance<any, true>,
  framework: PluginType,
  options: FeatureOptions | false
): Plugin[] | undefined

function resolvePlugin(
  unplugin: UnpluginInstance<any, false>,
  framework: PluginType,
  options: FeatureOptions | false
): Plugin | undefined

function resolvePlugin(
  unplugin: UnpluginInstance<any, boolean>,
  framework: PluginType,
  options: FeatureOptions | false
): Plugin | Plugin[] | undefined {
  if (!options) return
  return unplugin[framework!](options)
}

const name = 'unplugin-vue-macros'

export default createCombinePlugin<Options | undefined>(
  (userOptions = {}, meta) => {
    const options = resolveOptions(userOptions)

    const framework = meta.framework!
    const setupComponentPlugins = resolvePlugin(
      VueSetupComponent,
      framework,
      options.setupComponent
    )
    const namedTemplatePlugins = resolvePlugin(
      VueNamedTemplate,
      framework,
      options.namedTemplate
    )

    const plugins: OptionsPlugin[] = [
      resolvePlugin(VueSetupSFC, framework, options.setupSFC),
      setupComponentPlugins?.[0],
      resolvePlugin(VueSetupBlock, framework, options.setupBlock),
      namedTemplatePlugins?.[0],

      // props
      resolvePlugin(VueChainCall, framework, options.chainCall),
      resolvePlugin(VueDefineProps, framework, options.defineProps),
      resolvePlugin(VueDefinePropsRefs, framework, options.definePropsRefs),
      resolvePlugin(VueExportProps, framework, options.exportProps),

      // emits
      resolvePlugin(VueDefineEmit, framework, options.defineEmit),
      resolvePlugin(VueShortEmits, framework, options.shortEmits),

      // both props & emits
      resolvePlugin(VueDefineModels, framework, options.defineModels),

      // convert to runtime props & emits
      resolvePlugin(VueBetterDefine, framework, options.betterDefine),

      // runtime props
      resolvePlugin(VueDefineProp, framework, options.defineProp),

      resolvePlugin(VueDefineSlots, framework, options.defineSlots),
      resolvePlugin(VueExportExpose, framework, options.exportExpose),
      resolvePlugin(
        VueReactivityTransform,
        framework,
        options.reactivityTransform
      ),
      resolvePlugin(VueHoistStatic, framework, options.hoistStatic),
      resolvePlugin(VueDefineOptions, framework, options.defineOptions),
      resolvePlugin(VueJsxDirective, framework, options.jsxDirective),
      options.plugins.vue,
      options.plugins.vueJsx,
      resolvePlugin(VueDefineRender, framework, options.defineRender),
      setupComponentPlugins?.[1],
      namedTemplatePlugins?.[1],
      framework === 'vite'
        ? Devtools({ nuxtContext: options.nuxtContext })
        : undefined,
    ].filter(Boolean)

    return {
      name,
      plugins,
    }
  }
)
