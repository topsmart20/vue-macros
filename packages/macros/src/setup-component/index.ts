import {
  DEFINE_SETUP_COMPONENT,
  MagicString,
  babelParse,
  getLang,
  getTransformResult,
} from '@vue-macros/common'
import { walk } from 'estree-walker'
import { normalizePath } from '@rollup/pluginutils'
import {
  SETUP_COMPONENT_ID_REGEX,
  SETUP_COMPONENT_ID_SUFFIX,
  SETUP_COMPONENT_SUB_MODULE,
} from './constants'
import type { HmrContext, ModuleNode } from 'vite'
import type { CallExpression, Function, Node } from '@babel/types'

export * from './constants'

interface NodeContext {
  code: string
  body: string
  node: CallExpression
}
export type SetupComponentContext = Record<string, NodeContext[]>

export const scanSetupComponent = (code: string, id: string) => {
  const program = babelParse(code, getLang(id))
  const nodes: CallExpression[] = []
  walk(program, {
    enter(node: Node) {
      if (
        node.type === 'CallExpression' &&
        node.callee.type === 'Identifier' &&
        node.callee.name === DEFINE_SETUP_COMPONENT
      ) {
        nodes.push(node)
      }
    },
  })

  if (nodes.length === 0) return []

  // const s = new MagicString(code)
  const nodeContexts: NodeContext[] = nodes.map((node) => {
    if (
      !['FunctionExpression', 'ArrowFunctionExpression'].includes(
        node.arguments[0].type
      )
    )
      throw new SyntaxError(
        `${DEFINE_SETUP_COMPONENT}: Invalid setup component definition`
      )

    // const importName = `setupComponent_${i}`
    // s.overwrite(node.start!, node.end!, importName)
    // s.prepend(
    //   `import ${importName} from '${normalizedId}${SETUP_COMPONENT_ID_SUFFIX}${i}.vue'\n`
    // )
    const body: Node = (node.arguments[0] as Function).body
    let bodyStart = body.start!
    let bodyEnd = body.end!
    if (body.type === 'BlockStatement') {
      bodyStart++
      bodyEnd--
    }
    return {
      code: code.slice(node.start!, node.end!),
      body: code.slice(bodyStart, bodyEnd),
      node,
    }
  })
  return nodeContexts
}

export const transformSetupComponent = (
  code: string,
  id: string,
  ctx: SetupComponentContext
) => {
  const normalizedId = normalizePath(id)
  const s = new MagicString(code)

  const nodeContexts = scanSetupComponent(code, id)
  ctx[normalizedId] = nodeContexts

  for (const [i, { node }] of nodeContexts.entries()) {
    const importName = `setupComponent_${i}`
    s.overwrite(node.start!, node.end!, importName)
    s.prepend(
      `import ${importName} from '${normalizedId}${SETUP_COMPONENT_ID_SUFFIX}${i}.vue'\n`
    )
  }

  return getTransformResult(s, id)
}

export const loadSetupComponent = (
  virtualId: string,
  ctx: SetupComponentContext,
  root: string
) => {
  const index = +(SETUP_COMPONENT_ID_REGEX.exec(virtualId)?.[1] ?? -1)
  const id = virtualId.replace(SETUP_COMPONENT_ID_REGEX, '')
  const nodeCtx = ctx[id]?.[index] || ctx[root + id]?.[index]
  if (!nodeCtx) return

  const { body } = nodeCtx
  const lang = getLang(id)
  const code = `<script setup${lang ? ` lang="${lang}"` : ''}>
${body}
</script>`
  return code
}

export const hotUpdateSetupComponent = async (
  { file, modules, read }: HmrContext,
  ctx: SetupComponentContext
) => {
  const isSubModule = (id: string) => SETUP_COMPONENT_SUB_MODULE.test(id)
  const getSubModule = (module: ModuleNode): ModuleNode[] => {
    const importedModules = Array.from(module.importedModules)
    if (importedModules.length === 0) return []

    return importedModules
      .filter(({ id }) => id && isSubModule(id!))
      .flatMap((module) => {
        return [module, ...getSubModule(module)]
      })
  }

  const module = modules.find((mod) => mod.file === file)
  if (!module || !module.id) return

  const affectedModules = getSubModule(module)

  const normalizedId = normalizePath(file)
  const nodeContexts = scanSetupComponent(await read(), normalizedId)
  ctx[normalizedId] = nodeContexts

  return [...modules, ...affectedModules]
}
