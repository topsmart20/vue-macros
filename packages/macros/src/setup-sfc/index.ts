import { MagicString, babelParse, getLang } from '@vue-macros/common'

export const SETUP_SFC_REGEX = /\.setup\.[cm]?[jt]sx?$/

export const transfromSetupSFC = (code: string, id: string) => {
  const lang = getLang(id)
  const program = babelParse(code, lang)

  const s = new MagicString(code)
  for (const stmt of program.body) {
    if (stmt.type !== 'ExportDefaultDeclaration') continue

    if (
      !['FunctionExpression', 'ArrowFunctionExpression'].includes(
        stmt.declaration.type
      )
    )
      throw new SyntaxError(`Default export must be a function`)

    s.append(`defineOptions({
  render: ${s.sliceNode(stmt.declaration)}
});`)
    s.removeNode(stmt)
  }

  const attrs = `${lang ? ` lang="${lang}"` : ''}`
  s.prependRight(0, `<script setup${attrs}>\n`)
  s.append(`\n</script>`)

  return s
}
