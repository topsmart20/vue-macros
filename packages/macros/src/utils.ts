import { getPackageInfoSync } from 'local-pkg'

export const getVueVersion = (): 2 | 3 => {
  const vuePkg = getPackageInfoSync('vue')
  if (vuePkg) {
    return +vuePkg.version.slice(0, 1) as 2 | 3
  } else {
    return 3
  }
}
