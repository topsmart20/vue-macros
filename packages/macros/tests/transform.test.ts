/* eslint-disable unicorn/prefer-string-replace-all */

import { resolve } from 'path'
import { readFile } from 'fs/promises'
import { describe, expect, test } from 'vitest'
import glob from 'fast-glob'
import { transform } from '../src'

describe('transform', async () => {
  const root = resolve(__dirname, '..')
  const files = await glob('tests/fixtures/*.{vue,js,ts}', {
    cwd: root,
    onlyFiles: true,
  })

  for (const file of files) {
    test(file.replace(/\\/g, '/'), async () => {
      const filepath = resolve(root, file)
      const version = filepath.includes('vue2') ? 2 : 3
      try {
        const code = transform(
          await readFile(filepath, 'utf-8'),
          filepath,
          version
        )?.code
        expect(code).toMatchSnapshot()
      } catch (err) {
        expect(err).toMatchSnapshot()
      }
    })
  }
})
