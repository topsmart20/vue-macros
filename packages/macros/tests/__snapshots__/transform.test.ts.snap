// Vitest Snapshot v1

exports[`transform > tests/fixtures/basic.vue 1`] = `
"<script setup lang=\\"ts\\">import { emitHelper as __emitHelper } from 'unplugin-vue-macros/helper'
var { modelValue } = defineProps<{
  modelValue: string
visible: boolean
}>();
const _defineModel_emit = defineEmits<{
  (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
}>();



console.log(modelValue)
__emitHelper(_defineModel_emit, 'update:modelValue', 'newValue')
</script>

<template>
  <h1>123</h1>
</template>
"
`;

exports[`transform > tests/fixtures/decls.vue 1`] = `
"<script setup lang=\\"ts\\">let { modelValue } = defineProps<{
  modelValue: string
visible: boolean
}>();
const _defineModel_emit = defineEmits<{
  (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
}>();

let a = 'b'
</script>
"
`;

exports[`transform > tests/fixtures/error-rest.vue 1`] = `[SyntaxError: not supported]`;

exports[`transform > tests/fixtures/vue2.vue 1`] = `
"<script setup lang=\\"ts\\">import { emitHelper as __emitHelper } from 'unplugin-vue-macros/helper'



const {value, title,  id } = defineProps<{value: string
title: string
 id: string }>()

const emit = defineEmits<{(evt: 'input', value: string): void
(evt: 'update:title', value: string): void

  (evt: 'change'): void
}>()

{
  __emitHelper(emit, 'input', 'hello')
  __emitHelper(emit, 'update:title', 'world')
  emit('change')

  {
    let modelValue = 'world'
    modelValue = 'foo'
  }
}
</script>
"
`;

exports[`transform > tests/fixtures/with-define-emits.vue 1`] = `
"<script setup lang=\\"ts\\">const { modelValue } = defineProps<{
  modelValue: string
}>();



const emit = defineEmits<{(evt: 'update:modelValue', value: string): void

  (evt: 'change'): void
}>()

emit('change')
</script>
"
`;

exports[`transform > tests/fixtures/with-define-emits-assignment.vue 1`] = `
"<script setup lang=\\"ts\\">import { emitHelper as __emitHelper } from 'unplugin-vue-macros/helper'
let { modelValue, title } = defineProps<{
  modelValue: string
title: string
}>();
const _defineModel_emit = defineEmits<{
  (evt: 'update:modelValue', value: string): void
(evt: 'update:title', value: string): void
}>();



__emitHelper(_defineModel_emit, 'update:modelValue', __emitHelper(_defineModel_emit, 'update:title', '123'))
</script>
"
`;

exports[`transform > tests/fixtures/with-define-emits-decl.vue 1`] = `
"<script setup lang=\\"ts\\">import { emitHelper as __emitHelper } from 'unplugin-vue-macros/helper'
let { modelValue } = defineProps<{
  modelValue: string
}>();



const emit = defineEmits<{(evt: 'update:modelValue', value: string): void

  (evt: 'change'): void
}>()

{
  __emitHelper(emit, 'update:modelValue', 'hello')
  emit('change')

  {
    let modelValue = 'world'
    modelValue = 'foo'
  }
}
</script>
"
`;

exports[`transform > tests/fixtures/with-define-props.vue 1`] = `
"<script setup lang=\\"ts\\">const _defineModel_emit = defineEmits<{
  (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
}>();



defineProps<{modelValue: string
visible: boolean

  title?: string
}>()
</script>
"
`;

exports[`transform > tests/fixtures/with-define-props-decl.vue 1`] = `
"<script setup lang=\\"ts\\">const _defineModel_emit = defineEmits<{
  (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
}>();



const {modelValue,  title } = defineProps<{modelValue: string
visible: boolean

  title?: string
}>()
</script>
"
`;

exports[`transform > tests/fixtures/with-define-props-empty.vue 1`] = `
"<script setup lang=\\"ts\\">const _defineModel_emit = defineEmits<{
  (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
}>();



defineProps< {modelValue: string
visible: boolean
  } > ()
</script>
"
`;

exports[`transform > tests/fixtures/with-define-props-ref.vue 1`] = `
"<script setup lang=\\"ts\\">const _defineModel_emit = defineEmits<{
  (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
}>();

interface Props {modelValue: string
visible: boolean

  title?: string
}

defineProps<Props>()


</script>
"
`;

exports[`transform > tests/fixtures/with-define-props-rest.vue 1`] = `
"<script setup lang=\\"ts\\">const _defineModel_emit = defineEmits<{
  (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
}>();



const {modelValue,  title, ...props } = defineProps<{modelValue: string
visible: boolean

  title?: string
}>()

console.log(props)
</script>
"
`;

exports[`transform > tests/fixtures/without-decl.vue 1`] = `
"<script setup lang=\\"ts\\">defineProps<{
  modelValue: string
visible: boolean
}>();
const _defineModel_emit = defineEmits<{
  (evt: 'update:modelValue', value: string): void
(evt: 'update:visible', value: boolean): void
}>();


</script>
"
`;
